### PR TITLE
Add half precision base support

### PR DESCRIPTION
**Pull request by crtrott**
If possible we should move that in after the backend refactor since this file might be one of the per backend specializations.

### CODE REVIEW

**Comment by crtrott** [comment]
Forgot the license notice ...

File: core/src/Kokkos_Half.hpp
Comment (DavidPoliakoff) [inline core/src/Kokkos_Half.hpp:52]: Should this be IMPL? I feel like users might want to know whether their Kokkos has half-precision (though I'm open to other mechanisms)

Comment (crtrott) [reply core/src/Kokkos_Half.hpp:52]: half_t will always be defined, the way to figure out whether we use the fallback is half_is_float constexpr bool.

Comment (DavidPoliakoff) [reply core/src/Kokkos_Half.hpp:52]: @crtrott and I discussed in Slack. I was worried about a model where a code used different versions of Kokkos (including one before we make this change), and they might switch off functionality based on that macro. We're going with a model where codes use a version of Kokkos that evolves with the code, so this won't come up

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:169]: What about `[unsigned] short`?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (calewis) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:66]: Do we care that `double2half(val)` and `float2half(static_cast<float>(val))` don't have to return identical values (at least this is true for arm-gcc)?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:187]: Why "round-towards-zero (rz)" rather than 'round-to-nearest-even (rn)" mode?

File: core/unit_test/TestHalfConversion.hpp
Comment (dalg24) [inline core/unit_test/TestHalfConversion.hpp:52]: How did you come up with this relative tolerance?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:187]: because that is how the default rounding mode of floating point to int works in C and C++ or?

File: core/unit_test/TestHalfConversion.hpp
Comment (crtrott) [reply core/unit_test/TestHalfConversion.hpp:52]: wikipedia :-) This is the actual (or pretty close) value based on the number of bits for the mantisse.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:66]: Isn't that expected?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:156]: below this statement (i.e. in the else clause) comes the CUDA 11.1 stuff.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:48]: ah right need to move Kokkos_Macros.hpp up ...

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:169]: forgot short. I mostly did the exact ones due to the mess of long vs long long I guess we could do all 8
```
int
long int
long long int
unsigned int
unsigned long int
unsigned long long int
short
unsigned short
```

File: core/unit_test/TestHalfConversion.hpp
Comment (dalg24) [inline core/unit_test/TestHalfConversion.hpp:56]: Did you look into gtest support for fp comparison?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (calewis) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:66]: Users might be surprised if on one platform they get different results from `cast_to_half` than on another.

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:66]: that is the nature of half precision right now. To the degree vendors agree on conversions we should be good. But CUDA doesn't provide us the conversion calls on both sides so I rather call the fast one on the device and do the extra conversion (with potentially different result) on host. I think you argument above though that is just generally expected. static_cast<int>(val); doesn't necessarily give the same result as static_cast<int>(static_cast<float>(val)) either.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:76]: I would prefer if you wrote `[unsigned] short int, [unsigned] long long int`

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:145]:
```suggestion
    std::enable_if_t<std::is_same<T, float>::value, T>
```

File: core/src/Kokkos_Half.hpp
Comment (dalg24) [inline core/src/Kokkos_Half.hpp:101]: consider doing
```C++
namespace Impl {
template <class T>
struct is_in_my_list : std::integral_constant<bool, std::is_same<T, int>::value || ...> {};
}
template <class T>
KOKKOS_INLINE_FUNCTION std::enable_if_t<Impl::is_in_my_list<T>::value, T>
cast_from_half(half_t val) { return T(val); }

template <class T>
KOKKOS_INLINE_FUNCTION std::enable_if_t<Impl::is_in_my_list<T>::value, half_t>
cast_to_half(half_t val) { return half_t(val); }
```

Comment (crtrott) [reply core/src/Kokkos_Half.hpp:101]: I thought about it, actually but I liked that this listing was closer to the visuals in some way to the other variants.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:145]: ah right I am forgetting that we now got C++14

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:51]: right

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:76]: sure

File: core/src/Kokkos_Half.hpp
Comment (dalg24) [reply core/src/Kokkos_Half.hpp:101]: Christian made the case that `cast_to_half` is not templated in the CUDA backend implementation and that his intent was to mirror it.

**Comment by crtrott** [comment]
There are more changes coming to this. In particular for the CUDA version we need our own struct which does the right thing on the host and the device (i.e. overloaded math operators). Later we will add more versions: Intel only, CUDA on ARM (where we can actually use __fp16 on the host), HIP, etc.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:96]: ok we need to figure out why this is still an issue.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:116]: remove all the printf statements.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:313]: get rid of the #if 1

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:317]: don't need a tmp here.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:330]: don't need a tmp here.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:344]: don't need a tmp here.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:500]: lets change all of those.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:191]: should return reference.

**Comment by e10harvey** [comment]
I've attempted to address the CI build errors and feedback in https://github.com/crtrott/kokkos/pull/7.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (nliber) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:230]: These binary operators should be hidden friends, as in:

```
half_t friend operator+(half_t lhs, half_t rhs)
{
 #ifdef __CUDA_ARCH__
     lhs.val += rhs.val;
 #else
     lhs.val = __float2half(__half2float(lhs.val) + __half2float(rhs.val));
 #endif
     return lhs;
}
```

The benefits are symmetry and ADL.

It is less common to do those for unary operators, although Anthony Williams makes a case for it at [https://www.justsoftwaresolutions.co.uk/cplusplus/hidden-friends.html](https://www.justsoftwaresolutions.co.uk/cplusplus/hidden-friends.html).

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:172]: we had some issues with ambiguity before anything to consider in that direction? Or are you just saying we shouldn't define any of those constructors, and just write the explicit conversion constructors explicitly.

**Review by dalg24 — COMMENTED** [review]
Drive-by review.  I need more time.

File: core/unit_test/TestHalfConversion.hpp
Comment (crtrott) [reply core/unit_test/TestHalfConversion.hpp:56]: hm I am not sure that that works, since I want to explicitly cast b-base not b and base and then do the difference.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:121]: Good point. I will update accordingly. I used a regex to convert the cast_to_half forward decls above.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:58]: I think there is a better way to do this. I needed a compile time conditional since `using impl_type = T` was not possible in `Kokkos_Half.hpp`. Any suggestions?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:194]: This is a best faith effort to get half_t behaving like other precisions (ex: float). I observed that float can be implicitly cast to and from T. For half_t, we cannot have implicit conversion operators to T as the compiler will consider T and its operators in all expressions involving half_t; but, we cannot mark half_t's operators as higher precedence than T's.

**Comment by e10harvey** [comment]
@dalg24, @nliber: Thank you for your reviews.

@dalg24: Do you have any other feedback before I start addressing it?

@nliber: I've attempted to address your feedback in 
15104e6. Is there anything else you'd like addressed?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:58]: I am not quite following why using a type alias is not possible.
In any case you would probably need to comment to that effect if you absolutely have to use a macro and also make sure you `#undef` it when you don't need it any more.

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:58]: I have a suggestion, this is mostly about the test thing there we want to test interoperability with implementation types:
We can simply have a trait to do that however:
```c++
template<class T>
struct half_impl_type {
  using type = typename T::impl_type;
};

template<>
struct half_impl_type<float> {
  using type = float;
};
```
I would put that into the Impl namespace for now.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:121]: @dalg24: Thinking about this more now. I cannot overload based on return type only. This was the best way I could find to provide the forward decls for `cast_from_half`. Is there a better way to do this?

Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:121]: > ```c++
> std::enable_if_t<std::is_same<T, Foo>::value &&
>                  std::is_same<T, Bar>::value &&
>                  ... , T>
> ```

I don't think this expression will ever evaluate to true.

Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:121]: should be `&& -> ||`

Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:121]: I get a multiple definition error with `||`. Is there a better way to provide the forward decls than what is shown above?

**Review by nliber — APPROVED** [review]
Looks good, other than the question about short circuiting operators.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:371]: Thank you, @nilber. The unit [tests](https://github.com/kokkos/kokkos/pull/3439/files#diff-6980b46765a569d9ae64dffdbe6a0db53b9ed49c8b9157236055a90add5b47afR261-R265) were written to check for simple short circuiting. I will update the unit tests with some more complex expressions that rely on operator precedence as well.

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:371]: what kind of bugs? We could collect feedback on this. But since we don't have implicit conversion to bool (only from) statements like a&&b wouldn't work otherwise and you would need to do bool(a)&&bool(b) instead. So I am in favor of this.

Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:371]: In favor of keeping these logical operator overloads? Shall I update the unit tests with some more complex expressions involving these logical operators?

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:371]: I think I just merge it. Nevin did approve after all. And if its a real issue he shouldn't have :-)

**Comment by e10harvey** [comment]
@crtrott, @dalg24, @nliber: Is there anything else that should be addressed here before we squash and merge?

**Pull request merged**
